src/clients/UBAClient/UBAClientWithRefresh.ts

```diff
@@ -2,165 +2,1122 @@ import assert from "assert";
+      await forEachAsync(tokens, async (token) => {
+        let modifiedFlowsInBundle: Record<number, ModifiedUBAFlow[]>;
+
+        // We can skip the next step if we've already loaded flows for all chains from the cache:
+        if (
+          this.chainIdIndices.every((chainId) => {
+            return this.getBundleState(mostRecentBundleBlockRanges, token, chainId).loadedFromCache;
+          })
+        ) {
+          console.log(
+            `- Skipping validation for bundle ${bundleKeyForBlockRanges} for token ${token} because flows for all chains are already cached`
+          );
+          modifiedFlowsInBundle = Object.fromEntries(
+            this.chainIdIndices.map((chainId) => {
+              const cachedFlows = this.getBundleState(mostRecentBundleBlockRanges, token, chainId).flows;
+              return [chainId, cachedFlows];
+            })
+          );
+        } else {
+          // Validate flows, which should load them into memory. This is a big expensive function especially
+          // if we haven't loaded any data from the cache and a lot of time has passed since the UBA activation
+          // bundle!
+          modifiedFlowsInBundle = await this.validateFlowsInBundle(mostRecentBundleBlockRanges, token);
+        }
+
+        // Save into UBA client state
+        await forEachAsync(this.chainIdIndices, async (chainId) => {
+          if (!isDefined(newUbaClientState[chainId])) newUbaClientState[chainId] = {};
+          if (!isDefined(newUbaClientState[chainId][token])) newUbaClientState[chainId][token] = [];
+          const bundleState = this.getBundleState(mostRecentBundleBlockRanges, token, chainId);
+          newUbaClientState[chainId][token].push({
+            bundleBlockRanges: mostRecentBundleBlockRanges,
+            flows: bundleState.flows,
+            ubaConfig: bundleState.ubaConfig,
+            openingBalances: bundleState.openingBalances,
+          });
+
+          const redisKeyForBundle = this.getKeyForBundle(mostRecentBundleBlockRanges, token, chainId);
+          console.log(`- Storing new bundle state under key ${redisKeyForBundle}`);
+          // Note, we opt to store arrays as strings in redis rather than using the redis.json module because
+          // we don't plan to manipulate the data inside redis, so we really only want to optimize for writing
+          // and reading. The redis.json module is more performant for manipulating data while inside redis.
+          if (isDefined(this.cachingClient)) {
+            await this.cachingClient.set(
+              redisKeyForBundle,
+              modifiedFlowsInBundle[chainId]
+              // I don't think we want these keys to expire since we'll likely always need data from the beginning
+              // of the UBA activation block to validate even the latest bundles, because of the recursive nature
+              // of how we compute running balances as a function of all prior validated bundle history.
```

Comment:
Should we disable caching until we're some minimum number of bundles that have been executed since UBA genesis? Otherwise we risk needing to flush the cache if/when we detect issues.

Speaking of which, how would we flush the cache? `ProviderUtils` in `relayer-v2` has a cache namespace, so it's easy to implicitly flush it by just changing the namespace. This works well because we have automatic key expiry, but that's not the case here.

Reply:
yeah I think that's safe for now. We could add a hardcoded name-space to the beginning of each key so that we can more easily flush all UBA related keys: https://stackoverflow.com/questions/4006324/how-to-atomically-delete-keys-matching-a-pattern-using-redis